Use weasyprint for registration PDF export

abhinavohri opened the pull request:
This PR migrates our PDF generation to WeasyPrint and refactors the `RHRegistrationsExportPDFTable` and `RHRegistrationsExportPDFBook` classes into a single service. The export style is now controlled by a new `export_type` parameter.

The`prepare_participant_list_data` helper function returns dynamic and static headers separately, with their display order determined in the template. The function provides the participant data as a list of dictionaries.

abhinavohri committed:
migrate registration lists pdf from latex to weasy print

abhinavohri committed:
Fix ruff issues

abhinavohri committed:
Cleanup the templates and localize the title and column headers

abhinavohri committed:
cleanup handling of dynamic form items

abhinavohri committed:
Migrate participant book style to weasy print

abhinavohri committed:
Use single ednpoint for both table and book style

abhinavohri committed:
add export type to preserved args

ThiefMaster commented on indico/modules/events/registration/controllers/management/reglists.py:
```diff
@@ -451,29 +451,51 @@ def _process_args(self):
+            generation_date=generation_date
+        )
+        pdf = create_pdf(html, css, self.event)
+        return send_file(f'{filename}.pdf', pdf, 'application/pdf')
```
please move the `.pdf` into the two `filename` vars above. I like being able to search for the filename and find the relevant code

ThiefMaster commented on indico/modules/events/registration/controllers/management/reglists.py:
```diff
@@ -451,29 +451,51 @@ def _process_args(self):
-    """Export registration list to a PDF in book style."""
+    def _process_args(self):
+        super()._process_args()
+        self.export_type = request.view_args.get('export_type')
```
- Use webargs `@use_kwargs()` to get the export type.
- Do not use `.get()` on dicts when you know the value always exists (or expect it to always exist)

ThiefMaster commented on indico/modules/events/registration/templates/pdf/participants_book.html:
```diff
@@ -0,0 +1,66 @@
+        <div class="sub-section">
+            <h2>{%- trans %}Registration Details{% endtrans -%}</h2>
+            {% set static_items_order = [
+            "Registration Date",
+            "Registration State",
+            "Price",
+            "Checked in",
+            "Check-in Date",
+            "Tags"
```
missing indentation, and I don't particularly like having this logic inside the template nor using human-friendly column titles for this instead of internal names... it's just something that can easily break later on

abhinavohri committed:
Fetch export_type param from endpoint using use_kwargs

abhinavohri committed:
User inner names instead of human readable column names

abhinavohri committed:
replace if and else if statements with match

ThiefMaster commented on indico/modules/events/registration/util.py:
```diff
@@ -1168,3 +1168,47 @@ class CustomTicketCode:
+                        cell_data = format_datetime(registration.checked_in_dt)
+                case 'tags_present':
+                    if registration.tags:
+                        cell_data = ', '.join(sorted(t.title for t in registration.tags))
```
```suggestion
                case 'checked_in_date' if registration.checked_in_dt:
                    cell_data = format_datetime(registration.checked_in_dt)
                case 'tags_present' if registration.tags:
                    cell_data = ', '.join(sorted(t.title for t in registration.tags))
```

otherwise `cell_data` remains defined in case of no checkin / tags

abhinavohri committed:
Fix indentation and add checks to prevent rendering errors

abhinavohri committed:
Improve readability of match-case statements

abhinavohri committed:
Filter static column names based on export configuration

ThiefMaster committed:
Add changelog entry

ThiefMaster committed:
Show mod_data + other cleanups

ThiefMaster committed:
Include custom (plugin-based) data

ThiefMaster committed:
Refactor PDF data generation

- i18n for static column headers
- show correct data if two fields have the same title
- do not fail when data is missing in a registration

tomasr8 commented on indico/modules/events/registration/controllers/management/reglists.py:
```diff
@@ -501,31 +501,51 @@ def _process_args(self):
+            generation_date=format_date(now_utc(), format='full', timezone=self.event.tzinfo),
+        )
+        pdf = create_pdf(html, css, self.event)
+        return send_file(f'{filename}', pdf, 'application/pdf')
```
not needed?
```suggestion
        return send_file(filename, pdf, 'application/pdf')
```

ThiefMaster replied:
indeed, iirc it used to have he `.pdf` here instead of in the filename

tomasr8 commented on indico/modules/events/registration/fields/choices.py:
```diff
@@ -668,3 +673,8 @@ def render_placeholder(self, data, key=None):
+
+    def render_reglist_column(self, data):
+        content_dict = self.get_friendly_data(data)
+        text_val = f"{content_dict['choice']} ({content_dict['nights']} nights)"
```
Can we i18n this?

ThiefMaster replied:
yes

tomasr8 commented on indico/modules/events/registration/templates/pdf/participants_book.html:
```diff
@@ -0,0 +1,59 @@
+            </div>
+            <footer>
+                <div class="footer-generation-date">{{ generation_date }}</div>
+                <div class="footer-page-num">Page {{ loop.index }}</div>
```
missing i18n?

ThiefMaster replied:
yes

ThiefMaster committed:
Fix review comments
